File: WEEKLYCODE#4/counting_valleys/Python/counting_valleys.py
# Solution uses a two pointer technique, 
# we check if current step leads to height 0 and previous height was -ve.  
height = 0
prev_height = 0
c = 0
n = input()
s = input().strip()
for i in range(len(s)):
    if (s[i] == 'U'):
        height += 1
    elif s[i] == 'D':
        height -= 1
    if height == 0 and prev_height < 0:
        c += 1
    prev_height = height
#printing result
print (c)

File: WEEKLYCODE#4/counting_valleys/Java/counting_valleys.java
import java.util.*;

public class Solution {

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int n = input.nextInt();
        char[] str = input.next().toCharArray();
        
        int c = 0;
        int height = 0;
        
        for(char ch : str) {
            // Step up
            if(ch == 'U') {
                if(height == -1) {
                    c++;
                }
                height++;
            }
            // Step down
            else {
                height--;
            }
        }

        input.close();
        
        System.out.println(c);
    }
}